### DISCUSSION TITLE
The role of DPD function

### DISCUSSION THREAD

**Discussion by XYQ202** [strongswan/strongswan#1138] [Q&A]
Hello,team
       Can you introduce your DPD function in detail? My understanding is the function of reconnecting after an accidental disconnection and re-establishing an IPsec tunnel.

        In addition, I tried to establish an IPsec connection to the security gateway with the device today, but IKE_SA succeeded, but CHILD_SA did not succeed, the log is as follows:

```
root@au_i2c_nand_v2:/usr/etc# ipsec start
Starting strongSwan 5.9.5 IPsec [starter]...
root@au_i2c_nand_v2:/usr/etc#
root@au_i2c_nand_v2:/usr/etc# swanctl --load-all
loaded certificate from '/usr/etc/swanctl/x509/CrossfireCert.pem'
loaded certificate from '/usr/etc/swanctl/x509ca/CrossfireCACert.pem'
loaded RSA key from '/usr/etc/swanctl/rsa/CrossfireKey.pem'
no authorities found, 0 unloaded
no pools found, 0 unloaded
loaded connection 'OptusIPSEC'
successfully loaded 1 connections, 0 unloaded
root@au_i2c_nand_v2:/usr/etc# ipsec up OptusIPSEC
initiating IKE_SA OptusIPSEC[1] to 192.168.201.33
generating IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (336 bytes)
received packet: from 192.168.201.33[500] to 10.201.189.34[500] (265 bytes)
parsed IKE_SA_INIT response 0 [ SA KE No CERTREQ N(FRAG_SUP) ]
selected proposal: IKE:AES_CBC_256/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/MODP_1024
received cert request for "C=AU, O=Optus, CN=LTE-CA"
sending cert request for "C=AU, O=Optus, CN=LTE-CA"
authentication of '049162B0C182@mavenir.com' (myself) with RSA signature successful
sending end entity cert "C=AU, ST=NSW, L=SYDNEY, O=mavenir, CN=049162B0C182@mavenir.com"
establishing CHILD_SA OptusIPSEC{1}
generating IKE_AUTH request 1 [ IDi CERT N(INIT_CONTACT) CERTREQ IDr AUTH SA TSi TSr N(EAP_ONLY) N(MSG_ID_SYN_SUP) ]
splitting IKE message (1440 bytes) into 2 fragments
generating IKE_AUTH request 1 [ EF(1/2) ]
generating IKE_AUTH request 1 [ EF(2/2) ]
**sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)
retransmit 1 of request with message ID 1
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)
retransmit 2 of request with message ID 1
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)
retransmit 3 of request with message ID 1
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)
retransmit 4 of request with message ID 1
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)**
```

Can I ask what is the problem? Is the Local ID not verified at the security gateway?

**Comment by strongX509** [comment]
In the absence of ESP traffic, DPD checks if the connection is still active by periodically sending an IKEv2 Informational message, expecting a matching response. If no response is received after multiple retransmits, the connection is either cleared or restarted.

In your case the connection doesn't even complete but gets stuck int the IKE_AUTH message exchange.

* Either the peer doesn't like the digital signature or the IKEv2 identity of the client but just keeps silent about this problem, so that the client is doing retransmissions

or

* The MTU is lower than 1280 bytes and the IP fragments of the IKE_AUTH request don't arrive at the destination. In that case lowering the IKEv2 fragment size by explicitly setting the   **fragment_size** parameter in the **charon** section of **strongswan.conf** might help.

**Reply by XYQ202** [reply]
Thanks, My strongswan.conf uses the default configuration, I have not changed any operations, the default charon is only loaded with some plugins, how should you add the `fragment_size` , the gateway personnel also initially suspect that the ID or MTU length is too long Big

**Reply by strongX509** [reply]
In **/etc/strongswan.conf** add the following line in the **charon** section:
````
charon {
   ...
   fragment_size = 576
}
````
This is the rather low IPv4 fragment size recommended by the IKEv2 Fragmentation RFC 7883 which should always work.

**Reply by tobiasbrunner** [reply]
This can also be achieved by setting the `fragment_size` option to `0` (uses 576 for IPv4 and 1280 for IPv6).

**Reply by XYQ202** [reply]
Thanks,Why does the length of the MTU affect it?
In addition, in the log information above, is the IKE_SA stage established, and the CHILD_SA stage is not established?

**Reply by XYQ202** [reply]
Do C and O in the DN of the root certificate and the DN of the entity certificate need to be exactly the same?

**Comment by XYQ202** [comment]
I want to use the DPD function now. According to the instructions in the **swanctl.conf** file, I only need to add parameters such as **dpd_action** and **dpd_delay** in the correct place. In addition, can you explain the meanings of **clear**, **trap** and **restart** in dpd_action?

**Reply by strongX509** [reply]
For the use of **dpd_action** read the strongSwan documentation 
https://docs.strongswan.org/docs/5.9/swanctl/swanctlConf.html#_connections_conn_children

**Reply by XYQ202** [reply]
> 

Do I just need to add these items in swanctl.conf?

**Reply by Thermi** [reply]
Yes

**Comment by XYQ202** [comment]
Do C and O in the DN of the root certificate and the DN of the entity certificate need to be exactly the same?

**Reply by strongX509** [reply]
No, usually there is no such restriction, unless the CA certificate contains some name constraint extensions.

**Reply by XYQ202** [reply]
> No, usually there is no such restriction, unless the CA certificate contains some name constraint extensions.

Thank you for your support, I still want to ask me again what is the main reason for the failure to establish the tunnel in the log above? Can you explain in detail?

**sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)
retransmit 1 of request with message ID 1
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)
retransmit 2 of request with message ID 1
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)
retransmit 3 of request with message ID 1
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)
retransmit 4 of request with message ID 1
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (1252 bytes)
sending packet: from 10.201.189.34[500] to 192.168.201.33[500] (260 bytes)**

**Reply by XYQ202** [reply]
Ok，solved，thanks
